### PR TITLE
Finalize hypso resampling

### LINKED ISSUE #293: Incorporate solution to resample lake hypsography
In testing the new [`lake-temperature-process-model`](https://github.com/USGS-R/lake-temperature-process-models) GLM modeling pipeline, we discovered that, for some lakes, GLM model runs that initially failed [later succeeded](https://github.com/USGS-R/lake-temperature-process-models/issues/17#issuecomment-1024410663) if we decreased the resolution of the hypsography by resampling it to 1-meter intervals. For each lake, the hypsography is pulled into `lake-temperature-process-models` via [`7_config_merge/ou/nml_list.rds`](https://github.com/USGS-R/lake-temperature-model-prep/blob/main/7_config_merge.yml#L91) from this `lake-temperature-model-prep` pipeline.

While troubleshooting, we resampled the hypsography within the `lake-temperature-process-models` pipeline using `approx()`:
```r
#' @Title Resample incoming hypsography to 1-m intervals
#' @decription Resample the H and A values to 1-meter intervals; calc bsn_vals
#' @param H the vector of H (elevation) values from the lake-specific 
#' nml parameters loaded in from the nml_list.rds from lake-temperature-model-prep
#' @param A the vector of A (area) values from the lake-specific nml 
#' parameters loaded in from the nml_list.rds from lake-temperature-model-prep
#' @return a list containing H and A vectors and a bsn_vals scaler
resample_H_A <- function(H, A) {
  ha_df <- tibble(H = H,A = A) %>%
    arrange(H)
  
  # Resample hypso to 1-meter intervals
  # Add an additional row for the deepest (final) raw H value so we don’t 
  # end up with with a lake shallower than the lake depth param. 
  # Then remove duplicate rows if any exist, which they will if the
  # final H value from the raw H vector is an integer
  ha_df_resampled <- bind_rows(tibble(H=seq.int(floor(min(ha_df$H)), floor(max(ha_df$H))),
                  A=approx(ha_df$H, ha_df$A, xout=seq.int(floor(min(ha_df$H)), floor(max(ha_df$H))), rule=2)$y),
                  tibble(A=approx(H, A, xout=max(ha_df$H), rule=2)$y, H=max(ha_df$H))) %>%
                  distinct()

  c(ha_df_resampled, bsn_vals = nrow(ha_df_resampled))
}
```
However, since the hypsography is compiled here, it makes sense to include a more robust solution within this pipeline. 

Jordan notes:
> the method we use in lake-temp-prep should probably be a depth-specific resampling. I.e., if a lake is only 2m deep, we probably don't want to resample down to ints. But that's further down the road and not something to address during troubleshooting.

### PR DESCRIPTION
**Pull request by hcorson-dosch**
Fixes #293 

Add code to resample hypsography prior to creation of the `7_config_merge/out/nml_list.rds.ind` target.

This code is set up such that:
* Hypsography is only resampled for lakes with > 2 values in the H and A vectors (those lakes with additional bathymetric information beyond max depth)
* No duplicate elevation-area pairings are generated
* The resampled area values increase monotonically with elevation
* The elevation intervals to which the area values are resampled is determined based on lake depth, with an interval of 1m for deep lakes, 0.5m for medium-deep lakes, and 0.25m for shallow lakes. The current depth thresholds (8m, 5m, and 3m, respectively) were pulled from the values used to determine `max_layer_thick` [here](https://github.com/USGS-R/lake-temperature-model-prep/blob/main/7_config_merge/src/munge_lake_info.R#L149-L155) but can easily be adjusted
* The radaii rather than the area values are resampled with `approx()` and are then used to calculate the area values, per Lindsay's method used in [`lake-temperature-out`](https://github.com/USGS-R/lake-temperature-out/blob/main/2_process/src/calculate_toha.R#L302-L313)

### CODE REVIEW

**Comment by hcorson-dosch** [comment]
Marked this as draft b/c Lindsay and I are working to assess what impact this change will have on the TOHA calculations. Lindsay's done some digging and will add her notes once Github allows her to post.

**Comment by lindsayplatt** [comment]
I was just exploring what impact different resolutions would have on our thermal habitat code. I thought that maybe because we are just linearly interpolating, it wouldn't matter if we had 1m or 0.5m. I put together some test code and it looks like it might? Not sure what we should do about this? @jread-usgs any thoughts?

```r
# Set up temp profiles every 0.5m
depth_0.5m <- seq(0,5,0.5)
set.seed(9)
temp_0.5m <- sort(rnorm(11, mean=10, sd=3),decreasing = T)

# Get every other value in order to use every 1m
depth_1m <- depth_0.5m[c(T,F)]
temp_1m <- temp_0.5m[c(T,F)]

# See if we get the same value using both resolutions 
approx(temp_0.5m, depth_0.5m, xout = 12)$y
approx(temp_1m, depth_1m, xout = 12)$y

# OUTPUT
> approx(temp_0.5m, depth_0.5m, xout = 12)$y
[1] 0.8475811
> approx(temp_1m, depth_1m, xout = 12)$y
[1] 0.7261741
```

**Comment by hcorson-dosch** [comment]
For some context, here is [one spot](https://github.com/USGS-R/lake-temperature-out/blob/main/2_process/src/calculate_toha.R#L287-L288) where Lindsay is resampling hypso to get the areas at specific depths as part of calculating TOHA

And I plotted up examples of what my resampling code generates in a shallow, medium, and deep lake:
image
image
image

**Comment by hcorson-dosch** [comment]
@lindsayplatt - circling back to your mockup example. One key factor - the temperature predictions output from GLM will be at 0.5m intervals regardless of how we resample the hypsography.

In case it's helpful I ran some of your code from [here](A2_0.5m) with a dummy dataset and a 0.5m and 1m hypso:

```r
library(tidyverse)

resample_hypso <- function(hypso, new_depths) {
  
  # `rule = 2`: repeats top or bottom known hypso for any depths outside of known hypso
  hypso$radii <- sqrt(hypso$areas / pi)
  new_radii <- approx(hypso$depths, hypso$radii, xout=new_depths, rule = 2)$y
  
  # Now calculate area from new radii
  new_areas <- pi * new_radii^2
  
  # Create hypso at these new depths
  return(list(depths = new_depths, areas = new_areas))
}

get_tha <- function(wtr_df, hypso, wtr_lower, wtr_upper) {
  # run code pulled from function
  # https://github.com/USGS-R/lake-temperature-out/blob/main/2_process/src/calculate_toha.R#L162-L209
  z_wtr <- rLakeAnalyzer::get.offsets(wtr_df)
  z_max <- max(hypso$depths, z_wtr) # bottom could be in hypso or wtr
  z_surface <- 0 # always 0, might be in hypso, might not
  
  wtr_surface <- wtr_df[[which.min(z_wtr)]] # wtr_surface will be whatever the top-most wtr is
  wtr_bottom <- wtr_df[[which.max(z_wtr)]] # wtr_bottom will be whatever the bottom-most wtr is
  
  ##### Find exact depths of wtr thresholds 
  
  Z1_Z2 <- apply(wtr_df, MARGIN = 1, function(wtr_row) {
    # Using apply since these two actions need to be done per row
    wtr <- as.vector(t(wtr_row))
    if(length(unique(wtr)) == 1) {
      # In a well-mixed lake, it is possible for all wtr values to be the same
      # which causes approx to throw an error. Return NAs for the Zs and let
      # checks below determine if benth area is all (or partially) above or 
      # below the lake and set Zs based on that.
      Z1_Z2 <- c(NA,NA)
    } else {
      Z1_Z2 <- approx(wtr, z_wtr, xout=c(wtr_upper, wtr_lower))$y
    }
    return(Z1_Z2)
  })
  Z1 <- Z1_Z2[1,]
  Z2 <- Z1_Z2[2,]
  
  ##### Checks before calculating areas
  completely_below_lake <- wtr_bottom > wtr_upper # if THA is completely below lake, benthic area is 0 (too hot)
  completely_above_lake <- wtr_surface < wtr_lower # if THA is completely above lake, benthic area is 0 (too cold)
  benth_0 <- completely_below_lake | completely_above_lake
  
  # If completely above or completely below, the Zs need to reflect that
  Z1[completely_below_lake] <- z_max
  Z2[completely_below_lake] <- z_max
  Z1[completely_above_lake] <- z_surface
  Z2[completely_above_lake] <- z_surface
  
  # if THA extends above lake, use the surface as Z1 to calc THA
  extends_above_lake <- wtr_surface < wtr_upper & !completely_above_lake
  Z1[extends_above_lake] <- z_surface
  
  # if THA extends below lake, use the bottom as Z2 to calc THA
  extends_below_lake <- wtr_bottom > wtr_lower & !completely_below_lake
  Z2[extends_below_lake] <- z_max
  
  ##### Create hypso at these different depths
  A1 <- resample_hypso(hypso, Z1)[["areas"]]
  A2 <- resample_hypso(hypso, Z2)[["areas"]]
  
  return(list(A1=A1, A2=A2))
}

# set up inputs
wtr_df <- tibble(temp_0=sort(rnorm(11, mean=20, sd=3),decreasing = T), 
                 temp_0.5=sort(rnorm(11, mean=19, sd=3),decreasing = T),
                 temp_1=sort(rnorm(11, mean=18, sd=3),decreasing = T),
                 temp_1.5=sort(rnorm(11, mean=16, sd=3),decreasing = T),
                 temp_2=sort(rnorm(11, mean=13, sd=3),decreasing = T),
                 temp_2.5=sort(rnorm(11, mean=11, sd=3),decreasing = T),
                 temp_3=sort(rnorm(11, mean=10, sd=3),decreasing = T),
                 temp_3.5=sort(rnorm(11, mean=9, sd=3),decreasing = T),
                 temp_4=sort(rnorm(11, mean=9, sd=3),decreasing = T),
                 temp_4.5=sort(rnorm(11, mean=9, sd=3),decreasing = T),
                 temp_5=sort(rnorm(11, mean=9, sd=3),decreasing = T))

depth_0.5m <- seq(0,5,0.5)
set.seed(9)
depth_1m <- depth_0.5m[c(T,F)]

hypso_0.5m <- tibble(depths=depth_0.5m,
                     areas=sort(rnorm(11, mean=150, sd=3),decreasing = T))
interval<-1
hypso_1m <- tibble(depths=seq(floor(min(hypso_0.5m$depths)/interval)*interval, floor(max(hypso_0.5m$depths)/interval)*interval, by=interval),
                   areas=approx(hypso_0.5m$depths,hypso_0.5m$areas,xout=depths, rule=2)$y)

wtr_lower <- 11
wtr_upper <- 25

#### check created hypso using two different hypsos
results_0.5m <- get_tha(wtr_df,hypso_0.5m,wtr_lower,wtr_upper)
results_1m <- get_tha(wtr_df,hypso_1m,wtr_lower,wtr_upper)

A1_0.5m <- results_0.5m$A1
A1_1m <- results_1m$A1
A2_0.5m <- results_0.5m$A2
A2_1m <- results_1m$A2

> A1_0.5m
 [1] 153.7486 153.8327 153.8327 153.8327 153.8327 153.8327 153.8327 153.8327
 [9] 153.8327 153.8327 153.8327
> A1_1m
 [1] 153.4180 153.8327 153.8327 153.8327 153.8327 153.8327 153.8327 153.8327
 [9] 153.8327 153.8327 153.8327
> A2_0.5m
 [1] 146.4394 146.4394 147.6473 148.4014 149.0518 149.2334 149.2576 149.5047
 [9] 149.5550 149.8858 150.5272
> A2_1m
 [1] 146.4394 146.4394 147.4780 148.1243 148.8359 149.3197 149.3724 149.5303
 [9] 149.5623 150.3013 150.8115
```

**Comment by jread-usgs** [comment]
Cool/useful analysis. 

One thing I wanted to quick add is that 
> One key factor - the temperature predictions output from GLM will be at 0.5m intervals regardless of how we resample the hypsography.

we can actually ask `glmtools` to resample GLM at any depths we want, they don't have to be regular intervals. Just adding this so it isn't viewed as a constraint that we can't control.

**Comment by lindsayplatt** [comment]
Coming back to this and re-wrapping my head around the problem space. Based on these checks and short analyses, it looks like there are changes to the resulting lake areas, but they are pretty minimal ([the step that finds the `A1` and `A2`](https://github.com/USGS-R/lake-temperature-out/blob/main/2_process/src/calculate_toha.R#L138-L139) for the temperature range). I think the next stage of checking the potential impacts of resampling (and maybe more impacted by resampling) is the next part of the TOHA calcs, which is getting the cumulative benthic area of the habitat using [`calc_benthic_area_incl_hypso()`](https://github.com/USGS-R/lake-temperature-out/blob/main/2_process/src/calculate_toha.R#L323) and the `cum_ba` (calculated from hypso [here in `lake-temperature-out`](https://github.com/USGS-R/lake-temperature-out/blob/main/2_process/src/calculate_toha.R#L25-L29)).  

However, all that said, I may be missing a key issue. I have been using the hypso in the NML files. The GLM outputs return temp at depth intervals of 0.5, but the hypso intervals vary per lake (some are already more sparse than 1 m). 

```r
library(tidyverse)

# FUNCTIONS USED
calc_benthic_area <- function(Z1, Z2, A1, A2) {
  trap_length <- sqrt(A2 * pi) + sqrt(A1 * pi)
  trap_height <- sqrt( (Z1 - Z2)^2 + (sqrt(A2/pi) - sqrt(A1/pi))^2 )
  benthic_area <- trap_length * trap_height
  
  return(benthic_area)
}
resample_hypso <- function(hypso, new_depths) {
  
  # `rule = 2`: repeats top or bottom known hypso for any depths outside of known hypso
  hypso$radii <- sqrt(hypso$areas / pi)
  new_radii <- approx(hypso$depths, hypso$radii, xout=new_depths, rule = 2)$y
  
  # Now calculate area from new radii
  new_areas <- pi * new_radii^2
  
  # Create hypso at these new depths
  return(list(depths = new_depths, areas = new_areas))
}

all_nml <- readRDS('1_fetch/out/pb0_temp_projections_nml.rds')

# all_nml <- readRDS('1_prep/in/nml_list.rds')
all_morphometry <- purrr::map(all_nml, `[`, c("latitude", "longitude", "H", "A"))
site_morphometry <- all_morphometry[["nhdhr_105954667"]]

hypso <- data.frame(H = site_morphometry$H, A = site_morphometry$A) %>% 
  mutate(depths = max(H) - H, areas = A) %>% 
  arrange(depths) %>% 
  select(depths, areas)

# precalculate benthic areas for each known slice, rather than doing
# this for each time timestep

cum_benthic_area <- cumsum(calc_benthic_area(
  hypso$depths[-length(hypso$depths)],
  hypso$depths[-1],
  hypso$areas[-length(hypso$areas)],
  hypso$areas[-1]))

# Do that for resampled hypsos
hypso_resampled <- resample_hypso(hypso, seq(min(hypso$depths), max(hypso$depths), by = 1))
cum_benthic_area_resampled <- cumsum(calc_benthic_area(
  hypso_resampled$depths[-length(hypso_resampled$depths)],
  hypso_resampled$depths[-1],
  hypso_resampled$areas[-length(hypso_resampled$areas)],
  hypso_resampled$areas[-1]))

# Compare benthic areas before/after:
cum_benthic_area
cum_benthic_area_resampled

# OUTPUT VERY DIFFERENT
> cum_benthic_area
[1]  70943.62 195706.77 371622.21 433841.57 454889.24
> cum_benthic_area_resampled
[1]  47228.33 112530.16 192144.65 316994.65 393248.15 431131.78 453476.83
```

**Comment by hcorson-dosch** [comment]
> However, all that said, I may be missing a key issue. I have been using the hypso in the NML files. The GLM outputs return temp at depth intervals of 0.5, but the hypso intervals vary per lake (some are already more sparse than 1 m).

Wait now I'm confused -- what key issue do you feel like you're missing? Moving forward you would continue to use the hypso from the NML files, as you have been. It would just be the resampled hypso instead of the original hypso. Is your thought that to date (and potentially moving forward), the resolution of the temperature outputs would not match that of the hypsography?

**Comment by hcorson-dosch** [comment]
Okay Lindsay and I regrouped on this today and got some clarity. Lindsay ran a code snippet to compare the cumulative benthic area calculated with the original hypsography to that calculated with the resampled hypsography. In the process we realized that I probably should use the actual minimum elevation-area pair, just as I am [using the actual maximum elevation-area pair](https://github.com/hcorson-dosch/lake-temperature-model-prep/blob/finalize_hypso_resampling/7_config_merge/src/munge_lake_info.R#L97), so I'm going to make that change now

**Comment by hcorson-dosch** [comment]
Okay that did what we wanted it to - making the start and end values for the hypso identical to the original:
image
image
image

**Comment by lindsayplatt** [comment]
Building on my edit that the cumulative benthic areas are not as different as I originally thought once you line them up in depth order, we compared the THA/OHA calc outputs for two lakes by using the same Z1/Z1 (the depth at which the example THA/OHA range intersects our lake) and calculating the corresponding lake areas at those depths (A1 and A2), then calculating the benthic area. **The habitat area values came out to be the same regardless of whether we used the original hypso or the resampled hypso!!!** This matches our original assumption because ultimately, we are just linearly interpolating between known points to get the resampled hypso, so that final benthic area math should give us the same result.

```r
# FXNS needed
# resample_hypso(): https://github.com/USGS-R/lake-temperature-out/blob/main/2_process/src/calculate_toha.R#L302-L313
# calc_benthic_area(): https://github.com/USGS-R/lake-temperature-out/blob/main/2_process/src/calculate_toha.R#L315-L321
# calc_benthic_area_incl_hypso(): https://github.com/USGS-R/lake-temperature-out/blob/main/2_process/src/calculate_toha.R#L323-L377

Z1 <- 2 # Depth in lake for the top of the O/T range
Z2 <- 3 # Depth in lake for the bottom of the O/T range

# Current lake hypso
hypso_orig <- structure(list(
  depths = c(0, 1.524, 3.048, 4.572, 6.096, 7.3152), 
  areas = c(454739.276528491, 383881.802459821, 259155.913693587, 
            83253.6667468361, 21045.3337177386, 0)), 
  class = "data.frame", row.names = c(NA, -6L))

# Create a resampled version
max_depth_orig <- max(hypso_orig$depths)
hypso_resampled <- as_tibble(resample_hypso(hypso_orig, c(seq(0,max_depth_orig,by=0.5), max_depth_orig)))

# Calc benthic area using original hypso
A1_orig <- resample_hypso(hypso_orig, Z1)[["areas"]]
A2_orig <- resample_hypso(hypso_orig, Z2)[["areas"]]
cum_ba_orig <- cumsum(calc_benthic_area(
  hypso_orig$depths[-length(hypso_orig$depths)],
  hypso_orig$depths[-1],
  hypso_orig$areas[-length(hypso_orig$areas)],
  hypso_orig$areas[-1]))
habitat_orig <- calc_benthic_area_incl_hypso(Z1, Z2, A1_orig, A2_orig, hypso_orig, cum_ba_orig)

# Calc benthic area using resampled hypso
A1_resampled <- resample_hypso(hypso_resampled, Z1)[["areas"]]
A2_resampled <- resample_hypso(hypso_resampled, Z2)[["areas"]]
cum_ba_resampled <- cumsum(calc_benthic_area(
  hypso_resampled$depths[-length(hypso_resampled$depths)],
  hypso_resampled$depths[-1],
  hypso_resampled$areas[-length(hypso_resampled$areas)],
  hypso_resampled$areas[-1]))
habitat_resampled <- calc_benthic_area_incl_hypso(Z1, Z2, A1_resampled, A2_resampled, hypso_resampled, cum_ba_resampled)

# They are the same!!
> habitat_resampled
[1] 79614.48
> habitat_orig
[1] 79614.48
```

We are assuming the same Z1/Z2 as input here for both hypso and hypso_resampled because running those actual calculations using example GLM output would give the same Z1/Z2 depths because those were calculated based on the GLM output (currently the 0.5 m depth intervals), not the hypso themselves (though hypso is an input to GLM).

**Comment by lindsayplatt** [comment]
Adding one more double check since the example above was a site where we upsampled. Wanted to create an example that could be plotted and points actually seen, so I didn't use an actual lake example; I made one up.

**Upsampled (adding more depths to the hypso) results come out to be the exact same, while downsampled results (removing depths from the hypso) come out similar but not the exact same.** I think this is because when we remove points from the depth profile, the slope of the line between depths changes so when we calculate the area of the trapezoid, it is different.

image

image

```
# Upsampled output
> habitat_orig
[1] 40.20565
> habitat_resampled
[1] 40.20565

# Downsampled output
> habitat_orig
[1] 93.96417
> habitat_resampled
[1] 90.42993
```

```r
# Code used to get the plots above and results, need the functions noted in the previous comment.
# Upsample hypso example
hypso_orig <- tibble(depths = c(0,3,4,6,8), areas = c(144,120,80,36,0))
hypso_resampled <- resample_hypso(hypso_orig, seq(0,8,by=0.5))

# Downsample hypso example
set.seed(10)
hypso_orig <- tibble(depths = seq(0,6,0.5), areas = sort(sample(1:300, 13), decreasing = T))
hypso_resampled <- resample_hypso(hypso_orig, 0:6)

# Change what `hypso_orig` and `hypso_resampled` are and run the 
# following to create the plots and generate the habitat area values
habitat_orig <- calc_benthic_area_incl_hypso(
  Z1, Z2, 
  A1 = resample_hypso(hypso_orig, Z1)[["areas"]], 
  A2 = resample_hypso(hypso_orig, Z2)[["areas"]], 
  hypso = hypso_orig, 
  cum_ba = cumsum(calc_benthic_area(
    hypso_orig$depths[-length(hypso_orig$depths)],
    hypso_orig$depths[-1],
    hypso_orig$areas[-length(hypso_orig$areas)],
    hypso_orig$areas[-1])))

habitat_resampled <- calc_benthic_area_incl_hypso(
  Z1, Z2, 
  A1 = resample_hypso(hypso_resampled, Z1)[["areas"]], 
  A2 = resample_hypso(hypso_resampled, Z2)[["areas"]], 
  hypso = hypso_resampled, 
  cum_ba = cumsum(calc_benthic_area(
    hypso_resampled$depths[-length(hypso_resampled$depths)],
    hypso_resampled$depths[-1],
    hypso_resampled$areas[-length(hypso_resampled$areas)],
    hypso_resampled$areas[-1])))


plot(hypso_orig$areas, -hypso_orig$depths, type = "b", pch = 16)
points(hypso_resampled$areas, -hypso_resampled$depths, pch = 16, col = "red")
lines(hypso_resampled$areas, -hypso_resampled$depths, pch = 16, col = "red")
legend("top", col=c('black', 'red'), pch=16, legend = c('original', 'resampled'))
```

File: 7_config_merge/src/munge_lake_info.R [hunk 1]
```diff
@@ -66,6 +66,45 @@ munge_Kw <- function(out_ind, kw_varying_ind, ...){
...
+  lake_hypso$radii = sqrt(lake_hypso$A/pi)
+  hypso_resampled <- bind_rows(
+    # Add an additional row for the minimum (initial, deepest) raw H value so we retain the original area at the bottom of the lake
+    tibble(new_radii=approx(lake_hypso$H, lake_hypso$radii, xout=min(lake_hypso$H), rule=2)$y, H=min(lake_hypso$H)),
+    # Between min and max elevation (non-inclusive), interpolate based on specified interval
+    tibble(H=seq(floor((min(lake_hypso$H)+interval)/interval)*interval, floor((max(lake_hypso$H)-interval)/interval)*interval, by=interval),
+           new_radii=approx(lake_hypso$H, lake_hypso$radii, xout=H, rule=2)$y),
+    # Add an additional row for the maximum (final, surface) raw H value so we retain the original surface area
+    tibble(new_radii=approx(lake_hypso$H, lake_hypso$radii, xout=max(lake_hypso$H), rule=2)$y, H=max(lake_hypso$H))) %>%
+    mutate(A = pi * new_radii^2) %>%
```
Comment (lindsayplatt) [inline 7_config_merge/src/munge_lake_info.R:99]: There is a lot of repeated work happening here, so I think there is a way to simplify by using the fxn I shared as a little helper/utility. I think you can hide the `radii` stuff and focus this code on defining the depths that you want to have returned.

Use the fxn from `lake-temperature-out` with a slight adjustment to return a data.frame with A and H, not depths/areas:

```r
resample_hypso_util <- function(hypso, new_H) {
  
  # `rule = 2`: repeats top or bottom known hypso for any depths outside of known hypso
  hypso$radii <- sqrt(hypso$A / pi)
  new_radii <- approx(hypso$H, hypso$radii, xout=new_H, rule = 2)$y
  
  # Now calculate area from new radii
  new_A <- pi * new_radii^2
  
  # Create hypso at these new depths
  return(data.frame(H = new_H, A = new_A))
}
```

So start by just defining the new H's using your approach above but making a single vector of all these new Hs. Then, just pass those into the little helper fxn.

```r
new_interval_elevations <- c(
  min(lake_hypso$H),
  seq(from = floor((min(lake_hypso$H)+interval)/interval)*interval, 
      to = floor((max(lake_hypso$H)-interval)/interval)*interval, 
      by=interval),
  max(lake_hypso$H)
)

resample_hypso_util(lake_hypso, new_interval_elevations)
```

When I run a test of this, I get the same output as what you shared in the updated NML rds file for both site `nhdhr_105954667` and `nhdhr_106373442`

```r
# Load util fxn
resample_hypso_util <- function(hypso, new_H) {
  
  # `rule = 2`: repeats top or bottom known hypso for any depths outside of known hypso
  hypso$radii <- sqrt(hypso$A / pi)
  new_radii <- approx(hypso$H, hypso$radii, xout=new_H, rule = 2)$y
  
  # Now calculate area from new radii
  new_A <- pi * new_radii^2
  
  # Create hypso at these new depths
  return(data.frame(H = new_H, A = new_A))
}

# Morphometry before resampling for nhdhr_105954667
site1_morph <- list(
  H = c(446.6848, 447.904, 449.428, 450.952, 452.476, 454), 
  A = c(0, 21045.3337177386, 83253.6667468361, 259155.913693587, 
        383881.802459821, 454739.276528491)
)

# Morphometry before resampling for nhdhr_106373442
site2_morph <- list(
  H = c(398.5888, 398.8936, 399.1984, 399.5032, 399.808, 400.1128, 
        400.4176, 400.7224, 401.0272, 401.332, 401.6368, 401.9416, 
        402.2464, 402.5512, 402.856, 403.1608, 403.4656, 403.7704, 
        404.0752, 404.38, 404.6848, 404.9896, 405.2944, 405.5992, 
        405.904, 406.2088, 406.5136, 406.8184, 407.1232, 407.428, 
        407.7328, 408.0376, 408.3424, 408.6472, 408.952, 409.2568, 
        409.5616, 409.8664, 410.1712, 410.476, 410.7808, 411.0856, 
        411.3904, 411.6952, 412), 
  A = c(3699.73951147273, 39297.2331894265, 
        69870.0806389613, 99068.024891665, 159888.742671484, 191711.502118137, 
        203660.660810597, 213909.939186974, 224209.214043236, 235183.441377942, 
        247132.600070401, 260706.644359115, 278130.417598957, 306828.397052813, 
        350950.290551119, 386247.805349764, 417995.570076591, 447243.51080918, 
        475566.5166639, 504189.501397929, 533362.44741069, 563735.308940686, 
        596932.971584171, 662203.376073734, 709375.054845011, 743797.631245673, 
        775395.406532845, 869663.769355573, 929834.532896889, 964957.060015938, 
        986480.54460633, 1005079.23512346, 1022952.97668227, 1041351.68127986, 
        1062700.17819066, 1085323.58533851, 1104997.20017317, 1124445.83084834, 
        1143644.47912409, 1163868.05523748, 1184116.62959081, 1205665.11242115, 
        1228963.47204745, 1256686.52014355, 1300033.46820364)
)

# CHANGE ME to `site1_morph` to test the other site outputs
lake_hypso <- site2_morph 

lake_depth <- diff(range(lake_hypso$H))
interval <- dplyr::case_when(
    lake_depth >= 8 ~ 1,
    lake_depth >= 5 & lake_depth < 8 ~ 0.5,
    TRUE ~ 0.25
)

new_interval_elevations <- c(
  min(lake_hypso$H),
  seq(from = floor((min(lake_hypso$H)+interval)/interval)*interval, 
      to = floor((max(lake_hypso$H)-interval)/interval)*interval, 
      by=interval),
  max(lake_hypso$H)
)

resample_hypso_util(lake_hypso, new_interval_elevations)
```

```r
# Resampling results for site1 (nhdhr_105954667)
          H          A
1  446.6848      0.000
2  447.0000   1406.625
3  447.5000   9408.800
4  448.0000  23749.088
5  448.5000  40472.033
6  449.0000  61626.004
7  449.5000  89374.786
8  450.0000 137871.595
9  450.5000 196838.747
10 451.0000 262711.769
11 451.5000 301192.625
12 452.0000 342302.481
13 452.5000 384951.186
14 453.0000 407568.299
15 453.5000 430830.996
16 454.0000 454739.277

# Resampling results for site2 (nhdhr_106373442)
          H          A
1  398.5888    3699.74
2  399.0000   48977.03
3  400.0000  179598.17
4  401.0000  223280.28
5  402.0000  264001.41
6  403.0000  367415.54
7  404.0000  468497.89
8  405.0000  564852.39
9  406.0000  720128.83
10 407.0000  905271.21
11 408.0000 1002775.52
12 409.0000 1066247.11
13 410.0000 1132840.98
14 411.0000 1199593.82
15 412.0000 1300033.47
```

**Review by lindsayplatt — APPROVED** [review]
This is functional but I left a rather long comment about a way to significantly reduce the complexity of your resampling code.

File: 7_config_merge/src/munge_lake_info.R (hunk 1 above)
Comment (hcorson-dosch) [reply 7_config_merge/src/munge_lake_info.R]: Alright - suggested edits to reorganize and streamline this code have been made

**Review by lindsayplatt — APPROVED** [review]
🎉 what a large body of work! Feels good to be done

**Comment by jread-usgs** [comment]
Sorry! I hadn't hit submit on my review

**Pull request merged**